### PR TITLE
extend the cryptography cffi backend for pypy's brand new stdlib _ssl module

### PR DESCRIPTION
**Pull request by planrich**
Here are the changes I needed to apply to implement PyPy's new stdlib _ssl module. I hope these are useful and will be merged.

If some cffi methods are not added correctly (missing ifdef for version check, ...) just point it out, I'm happy to fix it.

### CODE REVIEW

**Comment by alex** [comment]
Jenkins, ok to test.

**Comment by alex** [comment]
@planrich thanks for submitting this! it looks like there's a handful of build issues that need to be sorted out on some of our supported platforms

**Comment by planrich** [comment]
I'll have a look at them

File: src/_cffi_src/openssl/ssl.py [hunk 1]
```diff
@@ -703,4 +722,8 @@
 static const long TLS_ST_BEFORE = 0;
 static const long TLS_ST_OK = 0;
 #endif
+
+Cryptography_STACK_OF_SSL_CIPHER * Cryptography_get_ssl_session_ciphers(const SSL_SESSION *s) {
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/ssl.py]: `SSL_SESSION` is entirely opaque in OpenSSL 1.1.0 so we need a different way to get this information. It looks like CPython uses `SSL_get_ciphers` in master (https://github.com/python/cpython/blob/4106e4237df5397efc785ce4daa257aecb5bff75/Modules/_ssl.c#L1650), although I'm not 100% sure this is what you're attempting to mirror.

File: src/_cffi_src/openssl/x509v3.py [hunk 2]
```diff
@@ -290,7 +299,17 @@
 DIST_POINT_NAME *DIST_POINT_NAME_new(void);
 void DIST_POINT_NAME_free(DIST_POINT_NAME *);
 
+void * X509_get_ext_d2i(const X509 *, int, int *, int *);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_VERSION_NUMBER >= 0x10001000L
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509v3.py]: You can obtain the CRLDP of an X509 object (if it exists) using `X509_get_ext`. You can find the index via `X509_get_ext_by_NID` and then use that index in `X509_get_ext`. The resulting output can be cast to a `Cryptography_STACK_OF_DIST_POINT` in cffi, so there's no need to have these functions.

File: src/_cffi_src/openssl/bio.py [hunk 3]
```diff
@@ -130,7 +132,14 @@
...
 long BIO_set_buffer_read_data(BIO *, void *, long);
 long BIO_set_nbio(BIO *, long);
+void BIO_set_retry_read(BIO *);
+void BIO_clear_retry_flags(BIO *);
+
+#define CRYPTO_LOCK_BIO ...
 """
 
 CUSTOMIZATIONS = """
+int * Cryptography_bio_references(const BIO * b) {
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/bio.py]: BIOs are opaque in 1.1.0 so this likely needs to be obtained in a different way. I see current CPython master doesn't consider them opaque though so maybe I'm wrong? cc @tiran

File: src/_cffi_src/openssl/x509_vfy.py [hunk 4]
```diff
@@ -15,17 +15,25 @@
...
 
 typedef ... Cryptography_STACK_OF_ASN1_OBJECT;
+typedef ... Cryptography_STACK_OF_X509_OBJECT;
 
-typedef ... X509_STORE;
+typedef ... X509_OBJECT;
+typedef struct {
+    Cryptography_STACK_OF_X509_OBJECT * objs;
+    ...;
+} X509_STORE;
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509_vfy.py]: This is opaque in 1.1.0. To get this data there is a function called `X509_STORE_get0_objects`, but we'll need to write a version that works with < 1.1.0 for our own use. Same style as we did in: https://github.com/pyca/cryptography/blob/master/src/_cffi_src/openssl/rsa.py#L90-L115

File: src/_cffi_src/openssl/x509_vfy.py [hunk 5]
```diff
@@ -133,7 +142,7 @@
 int X509_STORE_set_default_paths(X509_STORE *);
 int X509_STORE_set_flags(X509_STORE *, unsigned long);
 void X509_STORE_free(X509_STORE *);
-
+X509_VERIFY_PARAM *_X509_STORE_get0_param(X509_STORE *);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509_vfy.py]: OpenSSL 1.1.0 has this function (`X509_STORE_get0_param`) and we should mimic the way we implemented versions for < 1.1.0 like https://github.com/pyca/cryptography/blob/master/src/_cffi_src/openssl/rsa.py#L90-L115

File: src/_cffi_src/openssl/x509_vfy.py [hunk 6]
```diff
@@ -173,6 +182,8 @@
 void X509_VERIFY_PARAM_set_depth(X509_VERIFY_PARAM *, int);
 int X509_VERIFY_PARAM_get_depth(const X509_VERIFY_PARAM *);
 void X509_VERIFY_PARAM_free(X509_VERIFY_PARAM *);
+int Cryptography_X509_OBJECT_get_type(const X509_OBJECT *);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509_vfy.py]: Same as the other 1.1.0 functions

File: src/_cffi_src/openssl/x509_vfy.py [hunk 7]
```diff
@@ -173,6 +182,8 @@
...
+X509 * Cryptography_X509_OBJECT_data_x509(X509_OBJECT*);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509_vfy.py]: `X509_OBJECT_get0_X509` in 1.1.0

File: src/_cffi_src/openssl/ssl.py (hunk 1 above)
Comment (planrich) [inline src/_cffi_src/openssl/ssl.py]: Thanks, In 3.5.2 they did not use SSL_get_ciphers, changed that!

File: src/_cffi_src/openssl/err.py [hunk 8]
```diff
@@ -195,9 +196,34 @@
 static const int SSL_TLSEXT_ERR_ALERT_FATAL;
 static const int SSL_TLSEXT_ERR_NOACK;
 
-static const int SSL_AD_INTERNAL_ERROR;
-static const int SSL_AD_ACCESS_DENIED;
+static const int SSL_AD_CLOSE_NOTIFY;
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/err.py]: These will be added in a followup PR once #3253 merges.

Comment (reaperhulk) [inline src/_cffi_src/openssl/err.py]: #3256 covers this now.

File: src/_cffi_src/openssl/x509v3.py [hunk 9]
```diff
@@ -214,6 +219,10 @@
 int sk_GENERAL_NAME_push(struct stack_st_GENERAL_NAME *, GENERAL_NAME *);
 GENERAL_NAME *sk_GENERAL_NAME_value(struct stack_st_GENERAL_NAME *, int);
 
+void sk_GENERAL_NAME_pop_free(struct stack_st_GENERAL_NAME*,
+                              sk_GENERAL_NAME_freefunc);
+void GENERAL_NAME_free(struct GENERAL_NAME_st*);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/x509v3.py]: PR #3257 covers this now.

**Comment by planrich** [comment]
as we discussed on IRC, it seems that most of the functionality we need in PyPy already lived in other pull requests. I'll stop working on the PR just now and see if I can get everything when the oder PRs are merged!

**Comment by alex** [comment]
I believe I've merged all the other PRs, so you can merge/rebase this one now.

**Comment by reaperhulk** [comment]
Closing this since we should have landed everything needed in other PRs. If you need anything new feel free to open a new PR though!

**Pull request closed without merging**